packages/search/src/search.tsx

```diff
@@ -0,0 +1,417 @@
+// @ts-ignore
+import { Spinner, __experimentalInputControl as InputControl } from '@wordpress/components';
+import { close, search, Icon } from '@wordpress/icons';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+/**
+ * Internal variables
+ */
+const SEARCH_DEBOUNCE_MS = 300;
+
+const keyListener = (
+	methodToCall: ( e: MouseEvent< HTMLDivElement > | KeyboardEvent< HTMLDivElement > ) => void
+) => ( event: KeyboardEvent< HTMLDivElement > ) => {
+	switch ( event.key ) {
+		case ' ':
+		case 'Enter':
+			methodToCall( event );
+			break;
+	}
+};
+
+type Props = {
+	autoFocus: boolean;
+	className?: string;
+	compact: boolean;
+	delaySearch: boolean;
+	delayTimeout: number;
+	describedBy?: string;
+	dir?: 'ltr' | 'rlt';
+	disableAutocorrect: boolean;
+	disabled: boolean;
+	fitsContainer: boolean;
+	hideClose: boolean;
+	hideOpenIcon: boolean;
+	initialValue?: string;
+	inputLabel?: string;
+	isOpen: boolean;
+	maxLength?: number;
+	minLength?: number;
+	onBlur?: ( event: MouseEvent< HTMLInputElement > ) => void;
+	onClick?: () => void;
+	onKeyDown?: ( event: KeyboardEvent< HTMLInputElement > ) => void;
+	onSearch: ( search: string ) => void;
+	onSearchChange: ( search: string ) => void;
+	onSearchOpen: ( event?: MouseEvent< HTMLDivElement > | KeyboardEvent< HTMLDivElement > ) => void;
+	onSearchClose: ( event: MouseEvent< HTMLDivElement > | KeyboardEvent< HTMLDivElement > ) => void;
+	overlayStyling?: ( search: string ) => React.ReactNode;
+	placeholder?: string;
+	pinned: boolean;
+	searching: boolean;
+	value?: string;
+};
+
+type State = {
+	keyword: string;
+	isOpen: boolean;
+	hasFocus: boolean;
+};
+
+class Search extends React.Component< Props, State > {
+	static defaultProps = {
+		autoFocus: false,
+		compact: false,
+		delaySearch: false,
+		delayTimeout: SEARCH_DEBOUNCE_MS,
+		describedBy: null,
+		dir: undefined,
+		disableAutocorrect: false,
+		disabled: false,
+		fitsContainer: false,
+		hideClose: false,
+		hideOpenIcon: false,
+		isOpen: false,
+		onClick: noop,
+		onKeyDown: noop,
+		onSearchChange: noop,
+		onSearchOpen: noop,
+		onSearchClose: noop,
+		//undefined value for overlayStyling is an optimization that will
+		//disable overlay scrolling calculation when no overlay is provided.
+		overlayStyling: undefined,
+		pinned: false,
+		searching: false,
+	};
+
+	instanceId = uniqueId();
+	searchInput = React.createRef< HTMLInputElement >();
+	openIcon = React.createRef< HTMLDivElement >();
+	overlay = React.createRef< HTMLDivElement >();
+
+	// debounced `onSearch` will have a `cancel` function
+	onSearch: ( ( search: string ) => void ) & { cancel?: () => void } = this.props.delaySearch
+		? debounce( this.props.onSearch, this.props.delayTimeout )
+		: this.props.onSearch;
+
+	state = {
+		keyword: this.props.initialValue ?? '',
+		isOpen: this.props.isOpen,
+		hasFocus: this.props.autoFocus,
+	};
+
+	UNSAFE_componentWillReceiveProps( nextProps: Props ) {
```

Comment:
A new component shouldn't use the legacy method. The new `getDerivedStateFromProps` method is preferred. Or even better, do a functional component with hooks.

Reply:
This component cannot actually use `getDerivedStateFromProps` because it's being controlled externally and internally. It specifically needs to respond to changes in `isOpen`. If we turn this into an uncontrolled component then we can remove this logic and replace it with a `defaultIsOpen` as you suggested in another comment. Then consumers that need the component to respond to external `isOpen` may add a `key` prop that depends on their `isOpen` value to force the component to re-mount with the new default value.